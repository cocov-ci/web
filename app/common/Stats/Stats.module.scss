@import "styles/variables.module";

.stats {
  width: 173px;
  height: 100%;
  background: $colorGrayLight;
  z-index: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  position: relative;
  overflow: hidden;

  .value,
  .description {
    position: relative;
    z-index: 1;
  }

  .value {
    font-size: 40px;
    font-variation-settings: "wght" 700;
  }

  .description {
    margin-bottom: 0;
  }

  .chart {
    position: absolute;
    left: -5px;
    bottom: -5px;
  }
}
