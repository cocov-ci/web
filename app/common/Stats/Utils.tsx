interface GradientProps {
  ctx: CanvasRenderingContext2D
  isCoverage: boolean
}

// Arrays containing gradient start, gradient stop, and border
const chartColors = {
  coverage: ['rgba(17, 150, 255, 0.36)', 'rgba(255, 255, 255, 1)', '#1196FF'],
  issues: ['rgba(205,29,141,0.36)', '#E8E8E8', '#CD1D8D'],
}

export const generateGradient = ({ ctx, isCoverage }: GradientProps) => {
  const gradient = ctx.createLinearGradient(0, 35, 0, 80)
  const [colorStart, colorStop] = isCoverage
    ? chartColors.coverage
    : chartColors.issues

  gradient.addColorStop(0, colorStart)
  gradient.addColorStop(1, colorStop)

  return gradient
}

export const getBorderColor = (isCoverage: boolean) =>
  (isCoverage ? chartColors.coverage : chartColors.issues)[2]

export const options = {
  pointStyle: false,
  fill: true,
  borderWidth: 1,
  scales: {
    x: {
      display: false,
      beginAtZero: true,
    },
    y: {
      display: false,
      beginAtZero: true,
    },
  },
}
