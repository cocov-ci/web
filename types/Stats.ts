import { ChartProps, ChartTypeProps } from './Chart'

export interface StatProps {
  value?: number
  data?: ChartProps
}

export interface StatsProps {
  issues: StatProps | undefined
  coverage: StatProps | undefined
}

export interface StatsResponseProps {
  issues: number[]
  coverage: number[]
}

export interface StatsComponentProps {
  data: StatProps | undefined
  type: ChartTypeProps
}
