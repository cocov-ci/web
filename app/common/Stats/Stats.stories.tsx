import type { Meta, StoryObj } from '@storybook/react'

import Stats, { makeFakePoints } from './index'

const meta: Meta<typeof Stats> = {
  title: 'Common/Stats',
  component: Stats,
  tags: ['autodocs'],
}

export default meta

type Story = StoryObj<typeof Stats>

export const Coverage: Story = {
  args: {
    type: 'coverage',
    data: {
      data: makeFakePoints(),
      value: Math.ceil(Math.random() * 100),
    },
  },
}

export const Issues: Story = {
  args: {
    type: 'issues',
    data: {
      data: makeFakePoints(),
      value: Math.ceil(Math.random() * 10000),
    },
  },
}
