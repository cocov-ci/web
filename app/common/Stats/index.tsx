import Chart from 'app/common/Chart'
import Text from 'app/common/Text'
import { StatsComponentProps } from 'types/Stats'

import styles from './Stats.module.scss'

const Stats = ({ data: dataChart, type }: StatsComponentProps) => {
  const { data, value } = dataChart || {}

  const coverage = type === 'coverage'

  return (
    <div className={styles.stats}>
      {typeof value === 'number' && data && (
        <div className={styles.chart}>
          <Chart data={data} height={80} type={type} width={182} />
        </div>
      )}
      <Text className={styles.description} variant="description">
        {coverage ? 'Coverage' : 'Issues'}
      </Text>

      <Text className={styles.value}>
        {typeof value !== 'number' ? (
          <>&#8212;</>
        ) : coverage ? (
          `${value}%`
        ) : (
          value
        )}
      </Text>
    </div>
  )
}

export default Stats
