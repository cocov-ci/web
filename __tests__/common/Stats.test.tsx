import { render } from '@testing-library/react'

import Stats from 'app/common/Stats'

import '@testing-library/jest-dom'

describe('common/Stats', () => {
  it('renders Stats component without data', () => {
    const { container } = render(<Stats data={undefined} type="coverage" />)

    expect(container.querySelectorAll('p')[0]).toHaveTextContent('Coverage')
    expect(container.querySelectorAll('p')[1]).toHaveTextContent('â€”')
  })

  it('renders Stats component with coverage type', () => {
    const { container } = render(<Stats data={{ value: 20 }} type="coverage" />)

    expect(container.querySelectorAll('p')[0]).toHaveTextContent('Coverage')
    expect(container.querySelectorAll('p')[1]).toHaveTextContent('20%')
  })

  it('renders Stats component with issues type', () => {
    const { container } = render(<Stats data={{ value: 20 }} type="issues" />)

    expect(container.querySelectorAll('p')[0]).toHaveTextContent('Issues')
    expect(container.querySelectorAll('p')[1]).toHaveTextContent('20')
  })

  it('renders Stats snapshots', () => {
    const { container } = render(<Stats data={undefined} type="coverage" />)
    expect(container).toMatchSnapshot()
  })
})
