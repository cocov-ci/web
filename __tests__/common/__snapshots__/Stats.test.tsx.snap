// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/Stats renders Stats snapshots 1`] = `
<div>
  <div
    class="stats"
  >
    <p
      class="text description description"
    >
      Coverage
    </p>
    <p
      class="text value"
    >
      â€”
    </p>
  </div>
</div>
`;
